parser/prism/Helpers.cc

```diff
@@ -0,0 +1,783 @@
+#include "parser/prism/Helpers.h"
+#include "parser/prism/Parser.h"
+#include <cstdlib>
+#include <cstring>
+
+using namespace std;
+using namespace sorbet::parser::Prism;
+
+namespace sorbet::parser::Prism {
+
+namespace {
+// Parser instance for node creation. Must be set via PMK::setParser()
+// before using any PMK functions, and must remain alive during PMK usage.
+thread_local const Parser *prismParser = nullptr;
+
+const Parser *parser() {
+    ENFORCE(prismParser != nullptr, "PMK::setParser must be called before using PMK helpers");
+    return prismParser;
+}
+
+pm_arguments_node_t *createArgumentsNode(const vector<pm_node_t *> &args, core::LocOffsets loc) {
+    pm_arguments_node_t *arguments = PMK::allocateNode<pm_arguments_node_t>();
+    if (!arguments) {
+        return nullptr;
+    }
+
+    pm_node_t **argNodes = (pm_node_t **)calloc(args.size(), sizeof(pm_node_t *));
+    if (!argNodes) {
```

Comment:
If you switch to `new`, a failed allocation will raise an exception (instead of returning `nullptr`. You can replace this check with a `try`/`catch` that frees `arguments` and returns `nullptr`.

(That can be further simplified later with `unique_ptr` and friends, but don't worry about that for now)